### ISSUE TITLE
Snakemake wrappers require cluster nodes to contact github

### ISSUE THREAD

**Issue by lparsons** [snakemake/snakemake#1157]
**Snakemake version**

6.6.1 and 6.7.0

**Describe the bug**

When attempting to execute workflows with rules that use a snakemake wrapper, the jobs themselves (running on the cluster nodes) attempt to contact github. On many of the clusters I use, there is limited or no connectivity (my systems admins claim that most hpc systems "don't NAT").

**Logs**

```
WorkflowError:
Failed to open source file https://github.com/snakemake/snakemake-wrappers/raw/0.77.0/bio/cutadapt/se/environment.yaml
ConnectionError: HTTPSConnectionPool(host='github.com', port=443): Max retries exceeded with url: /snakemake/snakemake-wrappers/raw/0.77.0/bio/cutadapt/se/environment.yaml (Caused by NewConnectionError('
<urllib3.connection.HTTPSConnection object at 0x2affe8649760>: Failed to establish a new connection: [Errno -2] Name or service not known'))
```

**Minimal example**

Running any rule that utilizes a wrapper seems to produce this problem.

**Additional context**

I'm not certain this is actually a bug, or intentional behavior, or if it's even something the author(s) feel should be changed, however, it does prevent me from using wrappers in most (all) of my workflows since they should be portable to these clusters at my institution.

Additonal [Stack Overflow post](https://stackoverflow.com/questions/58413114/snakemake-wrappers-fails-to-open-environment-file-http-error-404-not-found).

**Comment by lparsons** [comment]
My apologies, updated to the correct error message.

**Comment by Angel-Popa** [comment]
Have you found a solution to this issue? I am currently having the same issue

**Comment by LakeTsai** [comment]
> 

You can download the yaml file [here](https://github.com/snakemake/snakemake-wrappers/raw/0.77.0/bio/cutadapt/se) and modify the setting of snakemake file like this:
```
wrapper:
"file://path2yaml"
```

**Comment by hazmup** [comment]
> You can download the yaml file [here](https://github.com/snakemake/snakemake-wrappers/raw/0.77.0/bio/cutadapt/se) and modify the setting of snakemake file like this:
```
wrapper:
"file://path2yaml"
```

This requires that I run the pipeline from a login node using the `--conda-create-envs-only` flag, then move each wrapper `yaml` file into a separate directory and change the `wrapper` directives to point to them. And even then, the jobs fail with no particular error.

**Comment by TBradley27** [comment]
I am also encountering this issue - this is a serious/severe issue for workflow manager functionality as GitHub will forbid connection requests (HTTP 403 error) if too many are made within a given short time interval. This is understandable as GitHub probably interprets this as a DoS attack

This, in effect, makes snakemake wrapper use non-functional for medium-highly parallel workflow executions and so this bug is breaking/severe

**Comment by m-jahn** [comment]
Hmm, but isn't a wrapper just downloaded one single time for the respective rule that is referencing it? Parallel execution should not be a problem in that case. Of course internet access is required for the initial download of wrapper files and envs, so without any connection it's tricky to run wrappers at all.
Maybe @fgvieira knows more?

**Comment by fgvieira** [comment]
I think what happens is that whenever a rule is executed, `snakemake` downloads the yaml from the `--wrapper-prefix` to check if the env is already installed and up to date. If it is, then it uses the local environment (as @m-jahn said) and if not, downloads it to `--conda-prefix`.

A workaround might be to clone the snakemake-wrappers repo locally, and use that path on `--wrapper-prefix` (like #2277).

**Comment by m-jahn** [comment]
wouldn't it be more economic to check for the presence of the env just one single time at the execution of the workflow, not for every execution of each rule? That feels like an unnecessary overhead (without knowing anything about snakemake wrappers internals)..

**Comment by fgvieira** [comment]
@m-jahn I agree that that would be the ideal scenario, but it would require changes in `snakemake`. Maybe adding a command-line option to skip checking if wrappers are installed/updated and use it internally when running individual rules (like e.g. `--wait-for-files`, `--runtime-source-cache-path` or `--allowed-rules`).

**Comment by m-jahn** [comment]
yes, an optional skipping of this check would circumvent the problem of hundreds or thousands of repetitive calls to github. That would be a good solution. Maybe something for the next hackathon.

**Comment by dlaehnemann** [comment]
@johanneskoester mentioned work on this on Discord. To properly cross-link this here, this is the open pull request that aims to address this #3643

**Comment by fgvieira** [comment]
@TBradley27 @lparsons @Angel-Popa Can you check if the PR fixed your issue?

You should be able to clone the repo locally and point to it when running `snakemake`:
```
snakemake --wrapper-prefix git+file://path/to/local/clone
```
No change to the original code required.
